sound/soc/sof/intel/cnl.c

```diff
@@ -198,13 +194,6 @@ static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
 		return 0;
 	}
 
-	/* ensure the DSP is in D0I0 before sending the message via mailbox */
-	ret = snd_sof_dsp_set_power_state(sdev, &target_state);
```

Comment:
@lyakh but the sof_ipc_tx_message_unlocked() also takes the ipc_lock and the snd_sof_dsp_send_msg() is called after that is also taken. How would I deal with that?

Reply:
> First question: do we really need to _set_ the power state for _each_ IPC? As far as I can see it really performs a bunch of IO potentially with some busy-waiting so it seems to take some time? I was told, that all DSP power state transitions only happen on request of the host, so doesn't the host know in which state the DSP is and shouldn't we skip setting the state, if it's already correct?

@lyakh the host knows only by checking if the requested state is the same as the state the DSP is in and thats the first thing snd_sof_dsp_set_power_state() does no?

Reply:
> > First question: do we really need to _set_ the power state for _each_ IPC? As far as I can see it really performs a bunch of IO potentially with some busy-waiting so it seems to take some time? I was told, that all DSP power state transitions only happen on request of the host, so doesn't the host know in which state the DSP is and shouldn't we skip setting the state, if it's already correct?
> 
> @lyakh the host knows only by checking if the requested state is the same as the state the DSP is in and thats the first thing snd_sof_dsp_set_power_state() does no?

@ranj063 Ok, sorry, I was looking at an outdated code. Now `hda_dsp_set_power_state()` does check for "the same state" before doing anything else. It still does look a bit out of place to me to have that check hidden in a specific implementation. You know, the best flow would be something like
```
int configure()
{
	return perform_config();
}

void action()
{
	perform_action();
}
```
The next best would be
```
int configure()
{
	return perform_config();
}

void action()
{
	if (configuration_needed)
		if (configure() < 0) {
			print("error");
			return;
		}
	perform_action();
}
```
and the case where you call `configure()` from `action()` on each call and only check inside `configure()` whether any configuration is needed, would be only my third choice... But it depends of course. And yes, I realise, that a test like `if (configuration_needed)` would be hardware-specific, so, it could also be a hook into respective implementations.

Reply:
> @lyakh but the sof_ipc_tx_message_unlocked() also takes the ipc_lock and the snd_sof_dsp_send_msg() is called after that is also taken. How would I deal with that?

@ranj063 you're right, we have only one instance of `ipc->msg` so you cannot call `.send_msg()` recursively.

Reply:
> Ok, sorry, I was looking at an outdated code. Now `hda_dsp_set_power_state()` does check for "the same state" before doing anything else. It still does look a bit out of place to me to have that check hidden in a specific implementation. You know, the best flow would be something like

@lyakh the main reason for performing the check inside hda_dsp_set_power_state() is that only the device state is platform agnostic. The substates are platform-specific and the common code does not have a notion of how the substates are defined for each platform. Does that make sense?